game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/SelectMainBattleCasualtiesTest.java

```diff
@@ -0,0 +1,526 @@
+        mock(SelectMainBattleCasualties.Select.class);
+    final CasualtyDetails expected = new CasualtyDetails(targetUnits, List.of(), false);
+    when(selectFunction.apply(any(), any(), anyCollection(), anyInt())).thenReturn(expected);
+
+    final CasualtyDetails details =
+        new SelectMainBattleCasualties(selectFunction).apply(delegateBridge, selectCasualties);
+
+    assertThat(details.getKilled().toArray(), is(targetUnits.toArray()));
+    // edit mode always sets auto calculated to true
+    assertThat(details.getAutoCalculated(), is(true));
+
+    verify(selectFunction)
+        .apply(
+            eq(delegateBridge),
+            eq(selectCasualties),
+            (Collection<Unit>) argThat(containsInAnyOrder(targetUnits.toArray())),
+            eq(0));
+  }
+
+  @Nested
+  class TransportCasualtiesNotRestricted {
+
+    @Test
+    void moreHitPointsThanHits() {
+      final List<Unit> targetUnits = List.of(givenAnyUnit(), givenAnyUnit());
+
+      targetUnits.forEach(
+          unit -> {
+            final UnitAttachment unitAttachment =
+                (UnitAttachment) unit.getType().getAttachment(UNIT_ATTACHMENT_NAME);
+            when(unitAttachment.getHitPoints()).thenReturn(2);
+          });
+
+      final FiringGroup firingGroup = new FiringGroup("", "", List.of(), targetUnits);
+
+      final FireRoundState fireRoundState = new FireRoundState();
+      fireRoundState.setDice(givenDiceRollWithHits(3));
+
+      final SelectCasualties selectCasualties =
+          new SelectCasualties(
+              battleState,
+              BattleState.Side.OFFENSE,
+              firingGroup,
+              fireRoundState,
+              (arg1, arg2) -> new CasualtyDetails());
+
+      final SelectMainBattleCasualties.Select selectFunction =
+          mock(SelectMainBattleCasualties.Select.class);
+      final CasualtyDetails expected = new CasualtyDetails();
+      when(selectFunction.apply(any(), any(), anyCollection(), anyInt())).thenReturn(expected);
+
+      final CasualtyDetails details =
+          new SelectMainBattleCasualties(selectFunction).apply(delegateBridge, selectCasualties);
+
+      assertThat(details, is(expected));
+
+      verify(selectFunction)
+          .apply(
+              eq(delegateBridge),
+              eq(selectCasualties),
+              (Collection<Unit>) argThat(containsInAnyOrder(targetUnits.toArray())),
+              eq(3));
+    }
+
+    @Test
+    void moreHitsThanHitPoints() {
+      final List<Unit> targetUnits = List.of(givenAnyUnit(), givenAnyUnit());
+
+      targetUnits.forEach(
+          unit -> {
+            final UnitAttachment unitAttachment =
+                (UnitAttachment) unit.getType().getAttachment(UNIT_ATTACHMENT_NAME);
+            when(unitAttachment.getHitPoints()).thenReturn(2);
+            when(unit.getHits()).thenReturn(1);
+          });
+
+      final FiringGroup firingGroup = new FiringGroup("", "", List.of(), targetUnits);
```

Comment:
Any thoughts to create a constructor that removes the no-op arguments? EG:
```
new FiringGroup(targetUnits);
```

Reply:
Since this constructor is only for tests, I'd rather not.

Reply:
Fair enough, it's a bit disturbing though that invalid values can be injected. It would seem to put the FiringGroup into an invalid state. It makes one wonder if there is a subset of behavior that could instead be a strategy object and we could test the strategy object directly. Just my 2 cents on seeing this code in test here, I'm not sure if it's practical to fix it or not.

Reply:
I've updated the tests to "correctly" create a FiringGroup.  The FiringGroup constructor is now private and the only way to create them is through `FiringGroup.groupBySuicideOnHit`.